pkg/apis/core/types.go

```diff
@@ -2138,6 +2138,33 @@ const (
 	PullIfNotPresent PullPolicy = "IfNotPresent"
 )
 
+// ResourceResizePolicy specifies how Kubernetes should handle resource resize.
+type ResourceResizePolicy string
+
+// These are the valid resource resize policy values:
+const (
+	// RestartNotRequired tells Kubernetes to resize the container in-place
+	// without restarting it, if possible. Kubernetes may however choose to
+	// restart the container if it is unable to actuate resize without a
+	// restart. For e.g. the runtime doesn't support restart-free resizing.
+	RestartNotRequired ResourceResizePolicy = "RestartNotRequired"
+	// 'RestartRequired' tells Kubernetes to resize the container in-place
+	// by stopping and starting the container when new resources are applied.
+	// This is needed for legacy applications. For e.g. java apps using the
+	// -xmxN flag which are unable to use resized memory without restarting.
+	RestartRequired ResourceResizePolicy = "RestartRequired"
+)
+
+// ContainerResizePolicy represents resource resize policy for a single container.
+type ContainerResizePolicy struct {
+	// Name of the resource type to which this resource resize policy applies.
+	// Supported values: cpu, memory.
+	ResourceName ResourceName
+	// Resource resize policy applicable to the specified resource name.
+	// If not specified, it defaults to RestartNotRequired.
+	Policy ResourceResizePolicy
```

Comment:
Realistically, I can only see us _maybe_ wanting something like RestartPod (if this container needs to be resized, restart the whole pod in-place (tear-down and re-create the pod sandbox) for VM pods that cannot be resized, or RestartEvict (if you cannot resize in place, evict me).

I don't recall the whole history in detail, but we did have https://github.com/kubernetes/enhancements/pull/686#discussion_r326999056 at one point, but delegated that to the VPA controller to implement in its policy set.

Regarding your suggestion, it will work well (assuming we will never need to have a policy beyond restart). While it didn't look so bad in my test pod yaml:

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: cpupolicypod
spec:
  restartPolicy: OnFailure
  terminationGracePeriodSeconds: 2
  containers:
  - name: stress
    image: skiibum/ubuntu-stress:18.10
    resizePolicy:
    - resourceName: "cpu"
      policy: RestartRequired
    resources:
      limits:
        cpu: "400m"
        memory: "500Mi"
      requests:
        cpu: "200m"
        memory: "500Mi"
```

Changing it to what you suggest looks cleaner and more explicit. (Pod.spec.containers[].resizePolicy[].restart = "NotRequired")

I'm still on the fence... my familiarity bias aside, the current API is generic and allows for a future extension.

Reply:
After sleeping over this, I still think what we have now is good. We have a concrete use case justifying RestartRequired (legacy applications) and a [plausible use case](https://github.com/kubernetes/enhancements/pull/686#discussion_r326999056) for a potential RestartPod policy. I'd rather keep the option of future extensibility open.

Reply:
I think I wasn't clear about what I meant.

`Pod.spec.containers[].resizePolicy[].policy = "RestartNotRequired"`  reduce to `....policy = "RestartNotRequired"`

There are 2 bits of information there:

1) what kind of policy ("restart")
2) what specific value ("not required")

If we ever have to add a different kind of policy (not "restart" but "termination" or "audit" or something else UNRELATED to restart policy), what will we call that field?  The generic word `policy` is used, so it will have to be `somethingPolicy = <value>`.  Since "something" is in the field name, we don't need it in the value.

so now we have:

```
    policy: RestartNotRequired
    somethingPolicy: TheValue
```

We have the same 2 bits of information, but organized differently.  In the first case its `policy: <policyType><policyValue>` and in the second it's `<policyType>Policy: <policyValue>`, which is far more future safe, so it make more sense to do it that way for both.

`restartPolicy: NotRequired`
`somethingPolicy: Optional`
`anotherPolicy: Purple`

Am I convincing?  It could be a followup PR after this merges but before 1.27 ?

Reply:
Yes, it makes sense. (And I'll do follow-up PR go in quickly after merge since we are on-topic and the context is fresh in our minds). Let me tweak around with the yaml to see how it looks.

Reply:
How does below look? `ResourceResizeRestartPolicy` is indeed a mouthful but quite precise .. I'm gonna to start calling it "aRRR Policy" ... now I just need my eye patch ;)

Do you think policy values of `NotRequired` and a more explicit `RestartContainer` (instead of `Required`) is a better choice? (say for [plausible extension ](https://github.com/kubernetes/enhancements/pull/686#discussion_r326999056)adding a value `RestartPod`)

```diff
diff --git a/staging/src/k8s.io/api/core/v1/types.go b/staging/src/k8s.io/api/core/v1/types.go
index 719aa7c7b6c..a41788398c0 100644
--- a/staging/src/k8s.io/api/core/v1/types.go
+++ b/staging/src/k8s.io/api/core/v1/types.go
@@ -2263,31 +2263,31 @@ const (
        PullIfNotPresent PullPolicy = "IfNotPresent"
 )
 
-// ResourceResizePolicy specifies how Kubernetes should handle resource resize.
-type ResourceResizePolicy string
+// ResourceResizeRestartPolicy specifies how to handle container resource resize.
+type ResourceResizeRestartPolicy string
 
-// These are the valid resource resize policy values:
+// These are the valid resource resize restart policy values:
 const (
-       // RestartNotRequired tells Kubernetes to resize the container in-place
+       // 'NotRequired' means Kubernetes will try to resize the container in-place
        // without restarting it, if possible. Kubernetes may however choose to
        // restart the container if it is unable to actuate resize without a
        // restart. For e.g. the runtime doesn't support restart-free resizing.
-       RestartNotRequired ResourceResizePolicy = "RestartNotRequired"
-       // 'RestartRequired' tells Kubernetes to resize the container in-place
+       NotRequired ResourceResizeRestartPolicy = "NotRequired"
+       // 'Required' means Kubernetes will resize the container in-place
        // by stopping and starting the container when new resources are applied.
        // This is needed for legacy applications. For e.g. java apps using the
        // -xmxN flag which are unable to use resized memory without restarting.
-       RestartRequired ResourceResizePolicy = "RestartRequired"
+       Required ResourceResizeRestartPolicy = "Required"
 )
 
-// ContainerResizePolicy represents resource resize policy for a single container.
+// ContainerResizePolicy represents resource resize policy for the container.
 type ContainerResizePolicy struct {
-       // Name of the resource type to which this resource resize policy applies.
+       // Name of the resource to which this resource resize policy applies.
        // Supported values: cpu, memory.
        ResourceName ResourceName `json:"resourceName" protobuf:"bytes,1,opt,name=resourceName,casttype=ResourceName"`
-       // Resource resize policy applicable to the specified resource name.
-       // If not specified, it defaults to RestartNotRequired.
-       Policy ResourceResizePolicy `json:"policy" protobuf:"bytes,2,opt,name=policy,casttype=ResourceResizePolicy"`
+       // Restart policy to apply when specified resource is resized.
+       // If not specified, it defaults to NotRequired.
+       RestartPolicy ResourceResizeRestartPolicy `json:"restartPolicy" protobuf:"bytes,2,opt,name=restartPolicy,casttype=ResourceResizeRestartPolicy"`
 }
 
 // PreemptionPolicy describes a policy for if/when to preempt a pod.
(END)
```

```yaml
...
kind: Pod
spec:
  restartPolicy: OnFailure
  terminationGracePeriodSeconds: 2
  containers:
  - name: stress
    image: skiibum/ubuntu-stress:18.10
    resizePolicy:
    - resourceName: "memory"
      restartPolicy: Required
    - resourceName: "cpu"
      restartPolicy: NotRequired
    resources:
...
```

Reply:
I'd keep the consts' *named* more verbosely (RestartNotRequired or even ResizeRestartNotRequired)

> Do you think a more explicit policy value of RestartContainer (instead of Required) is a better choice

So instead of "NotRequired" and "Required" you would do "NotRequired" and "RestartContainer" or similar?  I don't hate that!

Reply:
SG. `RestartNotRequired`, `RestartContainer` seem to sit well as values for `pod.spec.containers[].resizePolicy[resourceName].restartPolicy`,  and look fine in yaml as well.
```yaml
...
kind: Pod
spec:
  restartPolicy: OnFailure
  terminationGracePeriodSeconds: 2
  containers:
  - name: stress
    image: skiibum/ubuntu-stress:18.10
    resizePolicy:
    - resourceName: "cpu"
      restartPolicy: RestartNotRequired
    - resourceName: "memory"
      restartPolicy: RestartContainer
    resources:
...
```
This also leaves the door open for a future `RestartPod` policy which the kubelet is well equipped to enact.

I'll get this change ready to go soon after we merge the mothership.